File: json-tools/scripts/transform.py
#!/usr/bin/env python3
# Copyright 2026 Yevgeniy Zaremba
# SPDX-License-Identifier: Apache-2.0

"""Transform JSON to CSV/JSONL and CSV to JSON."""

from __future__ import annotations

import argparse
import csv
import io
import json
import sys
from typing import Any

from common import flatten_json, load_json, resolve_array, write_json


def parse_columns(raw: str | None) -> list[str]:
    if not raw:
        return []
    return [item.strip() for item in raw.split(",") if item.strip()]


def json_to_csv(data: Any, columns: list[str]) -> str:
    rows = data if isinstance(data, list) else [data]
    flattened = [flatten_json(row) if isinstance(row, (dict, list)) else {"value": row} for row in rows]
    if not columns:
        column_set: set[str] = set()
        for row in flattened:
            column_set.update(row.keys())
        columns = sorted(column_set)
    output = io.StringIO()
    writer = csv.DictWriter(output, fieldnames=columns, extrasaction="ignore")
    writer.writeheader()
    for row in flattened:
        writer.writerow({col: row.get(col) for col in columns})
    return output.getvalue()


def json_to_jsonl(data: Any) -> str:
    rows = data if isinstance(data, list) else [data]
    return "\n".join(json.dumps(row, ensure_ascii=False) for row in rows) + "\n"


def csv_to_json(path: str | None) -> Any:
    if not path or path == "-":
        text = sys.stdin.read()
    else:
        with open(path, "r", encoding="utf-8", newline="") as handle:
            text = handle.read()
    reader = csv.DictReader(io.StringIO(text))
    return list(reader)


def main() -> None:
    parser = argparse.ArgumentParser(description="Transform JSON<->CSV/JSONL.")
    parser.add_argument("input", nargs="?", default="-", help="Input file path or '-' for stdin.")
    parser.add_argument("--to", choices=["csv", "jsonl"], help="Convert JSON input to the target format.")
    parser.add_argument("--from-format", choices=["csv"], help="Convert non-JSON input into JSON.")
    parser.add_argument("--array-path", help="Path to array when converting JSON input.")
    parser.add_argument("--columns", help="Comma-separated columns for CSV output.")
    args = parser.parse_args()

    columns = parse_columns(args.columns)

    if args.from_format == "csv":
        result = csv_to_json(args.input)
        write_json(result)
        return

    data = load_json(args.input)
    if args.array_path:
        extracted = resolve_array(data, args.array_path)
        data = extracted if extracted else data

    if args.to == "csv":
        print(json_to_csv(data, columns), end="")
    elif args.to == "jsonl":
        print(json_to_jsonl(data), end="")
    else:
        # Default passthrough when no transform is requested.
        write_json(data)


if __name__ == "__main__":
    main()

File: json-tools/scripts-node/transform.mjs
#!/usr/bin/env node
// Copyright 2026 Yevgeniy Zaremba
// SPDX-License-Identifier: Apache-2.0

import fs from "node:fs";
import { parseArgs } from "node:util";
import { flattenJson, loadJson, resolveArray, writeJson } from "./common.mjs";

function parseColumns(raw) {
  if (!raw) {
    return [];
  }
  return raw.split(",").map((item) => item.trim()).filter(Boolean);
}

function csvEscape(value) {
  let text = "";
  if (value === true) {
    text = "True";
  } else if (value === false) {
    text = "False";
  } else if (value === null || value === undefined) {
    text = "";
  } else {
    text = String(value);
  }
  if (text.includes(",") || text.includes('"') || text.includes("\n")) {
    return `"${text.replaceAll('"', '""')}"`;
  }
  return text;
}

function jsonToCsv(data, columns) {
  const rows = Array.isArray(data) ? data : [data];
  const flattened = rows.map((row) => ((row && typeof row === "object")
    ? flattenJson(row)
    : { value: row }));

  let chosenColumns = columns;
  if (!chosenColumns.length) {
    const columnSet = new Set();
    for (const row of flattened) {
      for (const key of Object.keys(row)) {
        columnSet.add(key);
      }
    }
    chosenColumns = [...columnSet].sort();
  }

  const lines = [];
  lines.push(chosenColumns.map(csvEscape).join(","));
  for (const row of flattened) {
    lines.push(chosenColumns.map((col) => csvEscape(row[col])).join(","));
  }
  return `${lines.join("\r\n")}\r\n`;
}

function jsonWithSpaces(value) {
  if (value === null || typeof value !== "object") {
    return JSON.stringify(value);
  }
  if (Array.isArray(value)) {
    return `[${value.map((item) => jsonWithSpaces(item)).join(", ")}]`;
  }
  return `{${Object.entries(value)
    .map(([key, val]) => `${JSON.stringify(key)}: ${jsonWithSpaces(val)}`)
    .join(", ")}}`;
}

function jsonToJsonl(data) {
  const rows = Array.isArray(data) ? data : [data];
  return `${rows.map((row) => jsonWithSpaces(row)).join("\n")}\n`;
}

function parseCsvLine(line) {
  const out = [];
  let current = "";
  let inQuotes = false;
  for (let i = 0; i < line.length; i += 1) {
    const ch = line[i];
    if (inQuotes) {
      if (ch === '"' && line[i + 1] === '"') {
        current += '"';
        i += 1;
      } else if (ch === '"') {
        inQuotes = false;
      } else {
        current += ch;
      }
    } else if (ch === ",") {
      out.push(current);
      current = "";
    } else if (ch === '"') {
      inQuotes = true;
    } else {
      current += ch;
    }
  }
  out.push(current);
  return out;
}

function csvToJson(path) {
  const text = !path || path === "-" ? fs.readFileSync(0, "utf-8") : fs.readFileSync(path, "utf-8");
  const lines = text.split(/\r?\n/).filter((line) => line.length > 0);
  if (!lines.length) {
    return [];
  }
  const headers = parseCsvLine(lines[0]);
  const rows = [];
  for (const line of lines.slice(1)) {
    const values = parseCsvLine(line);
    const row = {};
    headers.forEach((header, idx) => {
      row[header] = values[idx] ?? "";
    });
    rows.push(row);
  }
  return rows;
}

function main() {
  const { values, positionals } = parseArgs({
    options: {
      to: { type: "string" },
      "from-format": { type: "string" },
      "array-path": { type: "string" },
      columns: { type: "string" },
    },
    allowPositionals: true,
  });

  if (values.to && !["csv", "jsonl"].includes(values.to)) {
    throw new Error("--to must be csv or jsonl");
  }
  if (values["from-format"] && values["from-format"] !== "csv") {
    throw new Error("--from-format currently supports only csv");
  }

  const input = positionals[0] ?? "-";
  const columns = parseColumns(values.columns);

  if (values["from-format"] === "csv") {
    const result = csvToJson(input);
    writeJson(result);
    return;
  }

  let data = loadJson(input);
  if (values["array-path"]) {
    const extracted = resolveArray(data, values["array-path"]);
    data = extracted.length ? extracted : data;
  }

  if (values.to === "csv") {
    process.stdout.write(jsonToCsv(data, columns));
  } else if (values.to === "jsonl") {
    process.stdout.write(jsonToJsonl(data));
  } else {
    writeJson(data);
  }
}

main();
